### ISSUE TITLE
Adding Correct Support QuantumLayer() and DynamicLayer()

### ISSUE THREAD

**Issue by Lemniscate-SHA-256** [Lemniscate-SHA-256/Neural-.nr#33]
Turn prototype layers QuantumLayer() and DynamicLayer() into real methods and add grammar.

**Comment by Lemniscate-SHA-256** [comment]
QuantumLayer using PennyLane (with an option to later swap in Qiskit if desired) and a DynamicLayer that makes decisions based on the input.

**Comment by Lemniscate-SHA-256** [comment]
The QuantumLayer will be implemented as a custom PyTorch module. It uses PennyLane to define a parameterized quantum circuit (a QNode) that processes classical data.

**Comment by Lemniscate-SHA-256** [comment]
QuantumLayer Module

```python

import torch
import torch.nn as nn
import pennylane as qml
from pennylane import numpy as pnp

class QuantumLayer(nn.Module):
    def __init__(self, n_qubits, n_layers, n_features):
        """
        n_qubits: Number of quantum bits to use.
        n_layers: Number of variational layers in the quantum circuit.
        n_features: Dimension of the input features per sample.
        """
        super(QuantumLayer, self).__init__()
        self.n_qubits = n_qubits
        self.n_layers = n_layers

        # Define a quantum device on the CPU or GPU via PennyLane's default.qubit device.
        self.dev = qml.device("default.qubit", wires=n_qubits)

        # The total number of parameters in the variational circuit.
        self.weight_shapes = {"weights": (n_layers, n_qubits, 3)}

        # Create a QNode that wraps our quantum circuit.
        @qml.qnode(self.dev, interface="torch")
        def circuit(inputs, weights):
            # Encode classical data into the quantum state.
            # For simplicity, we embed each feature onto a qubit using an RY gate.
            for i in range(self.n_qubits):
                if i < len(inputs):
                    qml.RY(inputs[i], wires=i)
                else:
                    qml.RY(0.0, wires=i)

            # Apply a variational circuit (multiple layers of rotations).
            for layer in range(self.n_layers):
                for qubit in range(self.n_qubits):
                    qml.RX(weights[layer, qubit, 0], wires=qubit)
                    qml.RY(weights[layer, qubit, 1], wires=qubit)
                    qml.RZ(weights[layer, qubit, 2], wires=qubit)
                # Entangle adjacent qubits with CNOTs
                for qubit in range(self.n_qubits - 1):
                    qml.CNOT(wires=[qubit, qubit + 1])
                # Optionally, add a CNOT from the last to the first qubit to create a ring entanglement.
                qml.CNOT(wires=[self.n_qubits - 1, 0])

            # Measure expectation values on each qubit.
            return [qml.expval(qml.PauliZ(wires=i)) for i in range(self.n_qubits)]

        # Register the quantum circuit as a torch parameter.
        self.weights = nn.Parameter(torch.randn(self.weight_shapes["weights"]))

        # Save the QNode function
        self.circuit = circuit

        # A final linear layer to map the quantum outputs to the desired feature size.
        self.fc = nn.Linear(n_qubits, n_features)

    def forward(self, x):
        # x is expected to be of shape (batch_size, n_features_in)
        # We will process each sample individually.
        outputs = []
        for sample in x:
            # For each sample, convert it to a 1D tensor (if not already).
            sample = sample.flatten()
            # Run the quantum circuit on the sample with current weights.
            q_out = self.circuit(sample, self.weights)
            q_out = torch.stack(q_out)  # Convert list to tensor.
            # Map the quantum outputs to desired features.
            outputs.append(self.fc(q_out))
        # Stack all outputs into a batch tensor.
        return torch.stack(outputs)

# Example usage:
if __name__ == "__main__":
    # Create a QuantumLayer with 4 qubits, 2 variational layers, mapping to 8 features.
    quantum_layer = QuantumLayer(n_qubits=4, n_layers=2, n_features=8)
    # Create a dummy input: batch of 3 samples, each with 4 features.
    dummy_input = torch.randn(3, 4)
    output = quantum_layer(dummy_input)
    print("QuantumLayer output shape:", output.shape)
```

**Comment by Lemniscate-SHA-256** [comment]
Quantum Circuit:
The QNode encodes the input features using RY rotations, applies variational rotations and entangling CNOTs, and measures the expectation values of the PauliZ operator on each qubit.
Parameter Handling:
The circuit parameters are stored as a torch Parameter (self.weights), so they are optimized during training.
Post-Processing:
A final linear layer maps the raw quantum outputs to the desired feature space.

**Comment by Lemniscate-SHA-256** [comment]
The DynamicLayer will dynamically decide which sub-module to use based on properties of the input (for example, the input’s norm or shape). This is implemented as a PyTorch module that contains two (or more) branches, selecting one at runtime.

**Comment by Lemniscate-SHA-256** [comment]
DynamicLayer Module

```python
class DynamicLayer(nn.Module):
    def __init__(self, threshold=1.0, branch1_units=64, branch2_units=32):
        """
        threshold: A value used to decide which branch to activate.
        branch1_units: Number of output features for branch 1.
        branch2_units: Number of output features for branch 2.
        """
        super(DynamicLayer, self).__init__()
        self.threshold = threshold

        # Define two branches as simple feedforward layers.
        self.branch1 = nn.Sequential(
            nn.Linear(128, branch1_units),
            nn.ReLU()
        )
        self.branch2 = nn.Sequential(
            nn.Linear(128, branch2_units),
            nn.ReLU()
        )
        # Final layer to map the branch outputs to a common dimension.
        # Here we choose the maximum of branch outputs as the final dimension.
        self.final_dim = max(branch1_units, branch2_units)
        self.final_layer = nn.Linear(self.final_dim, 128)

    def forward(self, x):
        """
        x: Expected shape (batch_size, 128)
        Decision logic: if the mean absolute value of the sample is greater than threshold,
        use branch1; otherwise, use branch2.
        """
        # Compute a simple statistic per sample
        branch_outputs = []
        for sample in x:
            if sample.abs().mean() > self.threshold:
                out = self.branch1(sample)
            else:
                out = self.branch2(sample)
            branch_outputs.append(out)
        # Stack outputs into a batch tensor
        branch_outputs = torch.stack(branch_outputs)
        # If branch outputs have different dimensions, we pad the smaller ones (for demonstration)
        if branch_outputs.shape[1] != self.final_dim:
            # A simple padding example (in practice, better design the branches to output the same dims)
            padded = torch.zeros(x.size(0), self.final_dim, device=x.device)
            padded[:, :branch_outputs.shape[1]] = branch_outputs
            branch_outputs = padded
        # Pass through a final mapping
        return self.final_layer(branch_outputs)

# Example usage:
if __name__ == "__main__":
    dynamic_layer = DynamicLayer(threshold=0.5, branch1_units=64, branch2_units=32)
    # Create a dummy input: batch of 5 samples, each with 128 features.
    dummy_input = torch.randn(5, 128)
    output = dynamic_layer(dummy_input)
    print("DynamicLayer output shape:", output.shape)
```

**Comment by Lemniscate-SHA-256** [comment]
Decision Logic:
For each sample in the batch, the DynamicLayer computes a simple statistic (mean absolute value) and then selects one of the two branches accordingly.
Branches:
Two separate sequential layers represent different processing pipelines.
Final Mapping:
A final linear layer maps the branch output to a consistent dimension.
Note:
In a production scenario, you might require the branches to always produce the same output dimension. Here, padding is provided as an example fallback.

**Comment by Lemniscate-SHA-256** [comment]
QuantumLayer and DynamicLayer are now fully integrated as explicit PyTorch modules using PennyLane and dynamic branch logic, respectively.
These examples provide a strong starting point for advanced functionalities in Neural.
